lib/data_services_api/sapint_converter.rb

```diff
@@ -0,0 +1,147 @@
+# frozen_string_literal: true
+
+module DataServicesApi
+  class SapiNTConverter
+    def initialize(dsapi_query)
+      @dsapi_query = JSON.parse(dsapi_query)
+    end
+
+    # Converts a DSAPI query to SAPINT query
+    def to_sapint_query
+      result = @dsapi_query.map do |key, value|
+        sapint_query(key, value)
+      end.compact.join('&')
+      sanitize_result(result)
+    end
+
+    private
+
+    def sapint_query(key, value)
+      case key
+      when '@sort'
+        sort(value)
+      when '@count'
+        count(value)
+      when '@limit'
+        limit(value)
+      when '@offset'
+        offset(value)
+      when '@and'
+        andd(value)
+      end
+    end
+
+    def sort(values)
+      values.map do |value|
+        sort_prop = value.key?('@up') ? "+#{value['@up']}" : "-#{value['@down']}"
+        "_sort=#{sort_prop}"
+      end
+    end
+
+    def count(value)
+      '_count=@id' if value == true
+    end
+
+    def limit(value)
+      "_limit=#{value}"
+    end
+
+    def offset(value)
+      "_offset=#{value}"
+    end
+
+    def andd(values)
+      values.map do |value|
+        attribute = remove_prefix(value.keys[0])
+        rel = value.values[0].keys[0]
+        json = value.values[0].values[0]
+        relation(rel, attribute, json)
+      end
+    end
+
+    def relation(rel, attribute, json)
+      case rel
+      when '@eq'
+        eq(attribute, json)
+      when '@oneof'
+        oneof(attribute, json)
+      when '@ge'
+        ge(attribute, json)
+      when '@gt'
+        gt(attribute, json)
+      when '@le'
+        le(attribute, json)
+      when '@lt'
+        lt(attribute, json)
+      when '@search'
+        search(attribute, json)
+      end
+    end
+
+    def eq(attribute, value)
+      return "#{attribute}=#{value}" unless value.is_a?(Hash)
+      return "#{attribute}=#{remove_prefix(value['@id'])}" if value.key?('@id')
+
+      "#{attribute}=#{value['@value']}"
+    end
+
+    def oneof(attribute, values)
+      values.map do |value|
+        eq(attribute, value)
+      end
+    end
+
+    def ge(attribute, value)
```

Comment:
There seems to be an opportunity to DRY -up `ge`, `gt`, etc as they are all very similar